src/dinit-iostream.cc

```diff
@@ -0,0 +1,740 @@
+}
+
+void istream::clear() noexcept
+{
+    eof_state = false;
+    string_failed = false;
+    io_error = 0;
+}
+
+getc_result istream::getc_nx() noexcept
+{
+    if (eof() || buffer_failure() || io_failure()) {
+        // We don't care about string_failure() in getc() function.
+        return getc_result(-1, false);
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(1);
+        if (r == 0) {
+            eof_state = true;
+            return getc_result(-1, false);
+        }
+        if (r < 0) {
+            io_error = errno;
+            return getc_result(-1, false);
+        }
+    }
+    char result = *buf->get_ptr(0);
+    buf->consume(1);
+    return getc_result(result, true);
+}
+
+getc_result istream::getc()
+{
+    getc_result r = getc_nx();
+    if (!r.success) {
+        throw_exception_on(io_states::eofbit | io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::get_line_nx(std::string &dest, char delim) noexcept
+{
+    if (!good()) {
+        return false;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(buf->get_size());
+        if (r == 0) {
+            eof_state = true;
+            return false;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return false;
+        }
+    }
+    dest.clear();
+
+    char *ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    char *eptr = static_cast<char *>(std::memchr(ptr, delim, len));
+
+    if (eptr == ptr) {
+        buf->consume(1);
+        return true; // An empty line
+    }
+    while (eptr == nullptr) {
+        // while no delimiter is found, append the entire buffer contents to the 'dest' string and refill the buffer.
+        try {
+            // dest.append() may throw length_error or bad_alloc
+            dest.append(ptr, len);
+        }
+        catch (const std::exception &e) {
+            string_failed = true;
+            return false;
+        }
+        buf->consume(len);
+
+        if (buf->get_length() == 0) {
+            int r = load_into_buf(buf->get_size());
+            if (r == 0) {
+                return true;
+            }
+            if (r < 0) {
+                io_error = errno;
+                return false;
+            }
+        }
+
+        ptr = buf->get_ptr(0);
+        len = buf->get_contiguous_length(ptr);
+        eptr = static_cast<char *>(std::memchr(ptr, delim, len));
+    }
+    unsigned index = eptr - ptr;
+    try {
+        // dest.append() may throw length_error or bad_alloc
+        dest.append(ptr, index);
+    }
+    catch (const std::exception &e) {
+        string_failed = true;
+        return false;
+    }
+    buf->consume(index + 1); // Consume the delimiter character as well
+    return true;
+}
+
+bool istream::get_line(std::string &dest, char delim)
+{
+    bool r = get_line_nx(dest, delim);
+    if (!r) {
+        // This will definitely throw an exception
```

Comment:
Add explanation

Reply:
Something like `// All possible exceptions are listed here so this will definitely throw an exception`?

Reply:
No, like in the examples I already gave above.

Reply:
```
Failed get_line_nx leaves an error state bit set, so this will throw:
```